File: tools/zetl-ast-js/src/dispatch.ts
// Declarative node-type dispatch (REQ-3218).
//
// Wraps `walk` with a table keyed by node type. Reserved keys:
//   - `Block`      — any block not otherwise covered
//   - `Inline`     — any inline not otherwise covered
//   - `_fallback`  — any node not otherwise covered
//   - `Blocks`     — sequence-level; receives each block-children array
//   - `Inlines`    — sequence-level; receives each inline-children array
//
// Visit order is depth-first pre-order: per-node handler fires on the
// node, then the sequence handler fires on its children array (if any),
// then each child is visited in turn. Handlers may mutate in place and
// return `undefined`/`null` to keep the existing node, or return a
// replacement node that will be swapped into the parent's children.

import type { AnyNode, BlockNode, InlineNode, ListItemNode, NodeOf, NodeType } from "./ast.ts";
import { isBlock, isInline } from "./ast.ts";
import type { Context } from "./context.ts";

export type NodeHandler<T extends AnyNode = AnyNode> = (
    node: T,
    ctx: Context,
) => AnyNode | undefined | null | void;

export type SequenceHandler<T extends AnyNode = AnyNode> = (
    children: T[],
    parent: AnyNode,
    ctx: Context,
) => T[] | undefined | void;

export type DispatchTable = {
    [K in NodeType]?: NodeHandler<NodeOf<K>>;
} & {
    Block?: NodeHandler<BlockNode>;
    Inline?: NodeHandler<InlineNode>;
    _fallback?: NodeHandler<AnyNode>;
    Blocks?: SequenceHandler<BlockNode | ListItemNode>;
    Inlines?: SequenceHandler<InlineNode>;
};

/**
 * Walk `root` and invoke handlers from `table` against each matching
 * node. Returns `root` (possibly a replacement produced by a root-level
 * handler).
 */
export function dispatch<T extends AnyNode>(root: T, ctx: Context, table: DispatchTable): T {
    return visit(root, ctx, table) as T;
}

function visit(node: AnyNode, ctx: Context, table: DispatchTable): AnyNode {
    // Per-node handler first — may replace the subtree before descent.
    const handler = resolveHandler(node, table);
    if (handler) {
        const r = handler(node as never, ctx);
        if (r != null && r !== (node as AnyNode)) node = r;
    }

    // Sequence-level handler on children (pre-descent).
    runSequence(node, ctx, table);

    // Descend — replacements bubble up via `visit()`'s return value.
    const arr = mutableChildren(node);
    if (arr) {
        for (let i = 0; i < arr.length; i++) {
            arr[i] = visit(arr[i]!, ctx, table) as never;
        }
    }
    return node;
}

function resolveHandler(node: AnyNode, table: DispatchTable): NodeHandler | undefined {
    const direct = (table as Record<string, NodeHandler | undefined>)[node.type];
    if (direct) return direct;
    if (isBlock(node) && table.Block) return table.Block as NodeHandler;
    if (isInline(node) && table.Inline) return table.Inline as NodeHandler;
    return table._fallback;
}

function runSequence(node: AnyNode, ctx: Context, table: DispatchTable): void {
    const blocksHandler = table.Blocks;
    const inlinesHandler = table.Inlines;
    if (!blocksHandler && !inlinesHandler) return;

    const kind = containerKind(node);
    if (kind === "block" && blocksHandler) {
        const arr = (node as { children: (BlockNode | ListItemNode)[] }).children;
        const next = blocksHandler(arr, node, ctx);
        if (next !== undefined) {
            (node as { children: (BlockNode | ListItemNode)[] }).children = next;
        }
    } else if (kind === "inline" && inlinesHandler) {
        const arr = (node as { children: InlineNode[] }).children;
        const next = inlinesHandler(arr, node, ctx);
        if (next !== undefined) {
            (node as { children: InlineNode[] }).children = next;
        }
    }
}

function containerKind(node: AnyNode): "block" | "inline" | null {
    switch (node.type) {
        case "Document":
        case "BlockQuote":
        case "ListItem":
        case "List":
            return "block";
        case "Paragraph":
        case "Heading":
        case "Emphasis":
        case "Strong":
        case "Link":
        case "Image":
            return "inline";
        default:
            return null;
    }
}

function mutableChildren(node: AnyNode): AnyNode[] | null {
    switch (node.type) {
        case "Document":
        case "Paragraph":
        case "Heading":
        case "Emphasis":
        case "Strong":
        case "Link":
        case "Image":
        case "BlockQuote":
        case "ListItem":
        case "List":
            return (node as { children: AnyNode[] }).children;
        default:
            return null;
    }
}

// ── onNode factory (REQ-3218) ───────────────────────────────────────────────

/**
 * Register a handler for a single node type. Python's helper uses
 * `@on_node("Wikilink")` as a decorator; JS has no decorator language
 * feature here, so this returns a `[type, handler]` tuple that's cheap
 * to compose into a dispatch table.
 *
 * ```ts
 * const links = onNode(Wikilink, (node, ctx) => { ... });
 * const notes = onNode(BlockQuote, (node, ctx) => { ... });
 * dispatch(ast, ctx, Object.fromEntries([links, notes]));
 * ```
 *
 * A three-argument form extends an existing table in-place:
 * ```ts
 * const table: DispatchTable = {};
 * onNode(table, Wikilink, handleLinks);
 * onNode(table, BlockQuote, handleQuotes);
 * ```
 */
export function onNode<T extends NodeType>(
    type: T,
    handler: NodeHandler<NodeOf<T>>,
): readonly [T, NodeHandler<NodeOf<T>>];
export function onNode<T extends NodeType>(
    table: DispatchTable,
    type: T,
    handler: NodeHandler<NodeOf<T>>,
): DispatchTable;
export function onNode(
    a: NodeType | DispatchTable,
    b: NodeType | NodeHandler,
    c?: NodeHandler,
): unknown {
    if (typeof a === "string") {
        return [a, b as NodeHandler] as const;
    }
    const table = a as DispatchTable;
    const type = b as NodeType;
    (table as Record<string, NodeHandler>)[type] = c as NodeHandler;
    return table;
}

File: tools/ztl-ast-py/src/ztl_ast/dispatch.py
"""
Declarative node-type dispatch (SPEC-032 REQ-3218).

Wraps a walk with a table keyed by node type. Reserved keys:

* ``Block``     — any block not otherwise covered
* ``Inline``    — any inline not otherwise covered
* ``_fallback`` — any node not otherwise covered
* ``Blocks``    — sequence-level; receives each block-children array
* ``Inlines``   — sequence-level; receives each inline-children array

Visit order is depth-first pre-order: the per-node handler fires on the
node, then the sequence handler fires on its children array (if any),
then each child is visited in turn. Handlers may mutate in place and
return ``None`` to keep the existing node, or return a replacement node
that is swapped into the parent's children.

Example::

    from ztl_ast import dispatch, on_node, Wikilink, BlockQuote

    @on_node(Wikilink)
    def link(node, ctx):
        node["alias"] = node["alias"] or node["target"]

    @on_node(BlockQuote)
    def quote(node, ctx):
        ...

    table = {}
    link(table)
    quote(table)
    dispatch(ast, ctx, table)
"""

from __future__ import annotations

from typing import Any, Callable, Dict, List, Optional

from .ast import BLOCK_TYPES, INLINE_TYPES, Node

NodeHandler = Callable[[Node, Any], Optional[Node]]
SequenceHandler = Callable[[List[Node], Node, Any], Optional[List[Node]]]
DispatchTable = Dict[str, Any]

_BLOCK_CONTAINERS = frozenset({"Document", "BlockQuote", "ListItem", "List"})
_INLINE_CONTAINERS = frozenset(
    {"Paragraph", "Heading", "Emphasis", "Strong", "Link", "Image"}
)
_CHILD_BEARING = _BLOCK_CONTAINERS | _INLINE_CONTAINERS


def dispatch(root: Node, ctx: Any, table: DispatchTable) -> Node:
    """
    Walk ``root`` and invoke handlers from ``table`` against each
    matching node. Returns ``root`` (possibly a replacement produced by
    a root-level handler).
    """
    return _visit(root, ctx, table)


def _visit(node: Node, ctx: Any, table: DispatchTable) -> Node:
    handler = _resolve_handler(node, table)
    if handler is not None:
        replacement = handler(node, ctx)
        if replacement is not None and replacement is not node:
            node = replacement

    _run_sequence(node, ctx, table)

    children = _mutable_children(node)
    if children is not None:
        for i, child in enumerate(children):
            children[i] = _visit(child, ctx, table)
    return node


def _resolve_handler(node: Node, table: DispatchTable) -> Optional[NodeHandler]:
    t = node.get("type")
    direct = table.get(t) if isinstance(t, str) else None
    if direct is not None:
        return direct
    if t in BLOCK_TYPES and "Block" in table:
        return table["Block"]
    if t in INLINE_TYPES and "Inline" in table:
        return table["Inline"]
    return table.get("_fallback")


def _run_sequence(node: Node, ctx: Any, table: DispatchTable) -> None:
    blocks = table.get("Blocks")
    inlines = table.get("Inlines")
    if blocks is None and inlines is None:
        return

    t = node.get("type")
    if t in _BLOCK_CONTAINERS and blocks is not None:
        kids = node.get("children")
        if isinstance(kids, list):
            nxt = blocks(kids, node, ctx)
            if nxt is not None:
                node["children"] = nxt  # type: ignore[index]
    elif t in _INLINE_CONTAINERS and inlines is not None:
        kids = node.get("children")
        if isinstance(kids, list):
            nxt = inlines(kids, node, ctx)
            if nxt is not None:
                node["children"] = nxt  # type: ignore[index]


def _mutable_children(node: Node) -> Optional[List[Node]]:
    if node.get("type") in _CHILD_BEARING:
        kids = node.get("children")
        if isinstance(kids, list):
            return kids
    return None


# ── on_node factory (REQ-3218) ──────────────────────────────────────────────


def on_node(node_type: str) -> Callable[[NodeHandler], NodeHandler]:
    """
    Declarative registration for a single node type.

    Used as a decorator, it attaches a ``__ztl_dispatch__`` tuple to
    the function and returns a callable that installs the handler into
    an existing dispatch table::

        @on_node("Wikilink")
        def link(node, ctx):
            ...

        table: DispatchTable = {}
        link(table)   # installs under "Wikilink"
        dispatch(ast, ctx, table)

    The decorator is intentionally thin — ``register.__ztl_dispatch__``
    exposes ``(type, original_handler)`` so tooling (coverage, audit)
    can enumerate hooks without re-parsing source.
    """

    def decorator(handler: NodeHandler) -> NodeHandler:
        def register(table: DispatchTable) -> DispatchTable:
            table[node_type] = handler
            return table

        register.__ztl_dispatch__ = (node_type, handler)  # type: ignore[attr-defined]
        register.__wrapped__ = handler  # type: ignore[attr-defined]
        register.__name__ = getattr(handler, "__name__", "register")
        register.__doc__ = handler.__doc__
        return register  # type: ignore[return-value]

    return decorator
